### PR TITLE
New Resources: `azurerm_(resource_group|subscription)_template_deployment`

### LINKED ISSUE #3408: Azurerm_template_deployment is not run, even if it was planned due to recreating a resource it's depending on unless parameters change
We have found out that when Data Factory is being recreated without a change (which was caused by https://github.com/terraform-providers/terraform-provider-azurerm/issues/3372 ), the ARM template is not actually deployed, even if it was planned, which is quite a problem in our case as the Data Factory is recreated, but is left empty.

I expect that this is how azurerm_template_deployment behaves in all similar cases where a resource it is depending on is recreated, but no parameter of the template is actually changed.

Please see the (obfuscated) output and notice that 1 change is planned, but 0 was applied. Also the ARM template is completely missing in the "progress" output.

### Output:
Terraform will perform the following actions:

-/+ ###MODULE_PATH###.azurerm_data_factory.factory (new resource required)
      id:                                           "/subscriptions/###SUB_ID###/resourceGroups/###RD_NAME###/providers/Microsoft.DataFactory/factories/###FACTORY_NAME###" => <computed> (forces new resource)
      identity.#:                                   "0" => <computed>
      location:                                     "###LOCATION###" => "###LOCATION###"
      name:                                         "###FACTORY_NAME###" => "###FACTORY_NAME###"
      resource_group_name:               "###RD_NAME###" => "###RD_NAME###" (forces new resource)
      tags.%:                                       "0" => <computed>

  ~ ###MODULE_PATH###.azurerm_template_deployment.pipelines
      parameters.%:                                 "" => <computed>
      parameters.backupBlobStorageConnectionString: ###CONNECTION_STRING###" => ""
      parameters.backupBlobStorageFolderPath:       "###BLOB_NAME###" => ""
      parameters.factoryId:                         "/subscriptions/###SUB_ID###/resourceGroups/###RD_NAME###/providers/Microsoft.DataFactory/factories/###FACTORY_NAME###" => ""
      parameters.factoryName:                       "###FACTORY_NAME###" => ""
      parameters.sourceCosmosDbConnectionString:    "###CONNECTION_STRING###" => ""
      parameters.triggerEnabled:                    "Stopped" => ""

Plan: 1 to add, 1 to change, 1 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

###MODULE_PATH###.azurerm_data_factory.factory: Destroying... (ID: /subscriptions/###SUB_ID###-.../factories/###FACTORY_NAME###)
###MODULE_PATH###.azurerm_data_factory.factory: Destruction complete after 2s
###MODULE_PATH###.azurerm_data_factory.factory: Creating...
  identity.#:          "" => "<computed>"
  location:            "" => "###LOCATION###"
  name:                "" => "###FACTORY_NAME###"
  resource_group_name: "" => "###RD_NAME###"
  tags.%:              "" => "<computed>"
###MODULE_PATH###.azurerm_data_factory.factory: Creation complete after 2s (ID: /subscriptions/###SUB_ID###-.../factories/###FACTORY_NAME###)

Apply complete! Resources: 1 added, 0 changed, 1 destroyed.

### Terraform (and AzureRM Provider) Version
Terraform v0.11.13
+ provider.azurerm v1.27.1
+ provider.fastly v0.6.0
+ provider.null v2.1.2
+ provider.random v2.1.2
+ provider.template v2.1.2

### LINKED ISSUE #5706: Output string value doesn't match what was specified in ARM template deploy
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.12.20
+ provider.azurerm v1.42.0

### Affected Resource(s)

* `azurerm_template_deployment`

### Terraform Configuration Files

```hcl
<some ARM template here>
	"outputs": {
		"LogicAppServicePrincipalId": {
			"type": "string",
			"value": "[reference(concat('Microsoft.Logic/workflows/',parameters('workflows_bh_keyvaulthandler_name')), '2019-05-01', 'Full').identity.principalId]"
		}
	}
```

I happen to be deploying a Logic App workflow definition and outputting the MSI principal ID.

### Expected Behavior

I should have an `outputs[]` collection with `LogicAppServicePrincipalId` in it. Terraform should not be changing the case of output values from an ARM template deployment.

### Actual Behavior

I got an error when trying to do
```hcl
output "LogicAppPrincipalId"{
	value=azurerm_template_deployment.logicapp.outputs["LogicAppServicePrincipalId"]
}
```

because the _actual_ value in `outputs` was `logicAppServicePrincipalId` (note the lowercase first character).

### Steps to Reproduce

1. `terraform apply`

### LINKED ISSUE #6561: Support for subscription scoped deployments in azurerm_template_deployment
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Right now azurerm_template_deployment requires a resource group name.
Arm templates are not necessary deployed at the resource group level, but also at the subscription level:

    budgets
    deployments - for nested templates that deploy to resource groups.
    eventSubscriptions
    peerAsns
    policyAssignments
    policyDefinitions
    policySetDefinitions
    remediations
    resourceGroups
    roleAssignments
    roleDefinitions
    scopeAssignments
    supportPlanTypes
    tags

 Support for this should be added.

### New or Affected Resource(s)

* azurerm_template_deployment

### References

* https://docs.microsoft.com/nb-no/azure/azure-resource-manager/templates/deploy-to-subscription?view=azure-cli-latest#supported-resources

### PR DESCRIPTION
**Pull request by tombuildsstuff**
This PR introduces two new resources:

- `azurerm_resource_group_template_deployment`
- `azurerm_subscription_template_deployment`

Which allow deploying an ARM Template Deployment at both the Resource Group and Subscription level. Whilst we currently have a Template Deployment resource in the form of the `azurerm_template_deployment` field - there's several limitations to this:

- we don't track external changes to the Template Body, we instead rely on the local template content
- we don't support import
- we don't support parameters or outputs of non-string values
- we don't clean up the resources deployed by this Template Deployment during the deletion of this resource

As such, whilst this Template Deployment resource is usable, it's not ideal - and ultimately this leads to us recommending users deploy an ARM Template into it's own Resource Group, to allow the Resource Group deletion to clear up items which have been deployed.

---

Instead this PR introduces two new resources  `azurerm_resource_group_template_deployment` and `azurerm_subscription_template_deployment` - which can be used to deploy an ARM Template at either a Resource Group or a Subscription level.

These new resources support specifying Parameters (and retrieving Outputs) as JSON blobs - so we support all of the available types returned from Azure (Strings, Ints, Bools, Objects etc) via the `jsondecode` interpolation function.

In addition the Resource Group Template Deployment now attempts to delete the resources provisioned by this Template Deployment when it's being deleted (there's a feature-toggle for this in the provider block). It's worth noting this is best-effort for the moment and likely requires further tweaking - but seems to work with a nested deployment.

Unfortunately we're unable to do the same thing for Subscription Template Deployments due to a lack of support in the Azure API for a generic delete method at this level.

Whilst the existing `azurerm_template_deployment` resource won't be going away anytime soon - we can supersede it in this release, deprecate it in 3.0 and then remove it in the future once there's been sufficient time to switch-over.

This PR also updates the Resources package to `2020-06-01`, so that we can gain this new functionality.

Fixes #3408
Fixes #5706
Fixes #6561

### CODE REVIEW

File: website/docs/r/resource_group_template_deployment.html.markdown [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  // NOTE: whilst we show an inline template here, we recommend
+  // sourcing this from a file for readability/editor support
+}
+```
+
+## Arguments Reference
+
+The following arguments are supported:
+
+* `deployment_mode` - (Required) The Deployment Mode for this Resource Group Template Deployment. Possible values are `Complete` (where resources in the Resource Group not specified in the ARM Template will be deployed) and `Incremental` (where resources are additive only).
```
Comment (jackofallops) [inline website/docs/r/resource_group_template_deployment.html.markdown:55]: Typo?
```suggestion
* `deployment_mode` - (Required) The Deployment Mode for this Resource Group Template Deployment. Possible values are `Complete` (where resources in the Resource Group not specified in the ARM Template will be destroyed) and `Incremental` (where resources are additive only).
```

**Review by jackofallops — APPROVED** [review]
Thanks @tombuildsstuff - one question and one minor typo(?), but otherwise LGTM 👍

File: azurerm/internal/services/resource/template_deployment_resource_group_resource.go [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Retrieving Template Deployment %q (Resource Group %q)..", id.Name, id.ResourceGroup)
+	template, err := client.Get(ctx, id.ResourceGroup, id.Name)
+	if err != nil {
+		return fmt.Errorf("retrieving Template Deployment %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
+	}
+	if template.Properties == nil {
```
[the remark this answers is not in the document]
Comment (tombuildsstuff) [reply azurerm/internal/services/resource/template_deployment_resource_group_resource.go:169]: we're not using the ID anywhere, so to be honest it felt superfluous - there is a CheckExists method in the SDK we could use instead, but we may as well reuse the Get here as we do elsewhere

**Comment by tombuildsstuff** [comment]
Tests pass:

Screenshot 2020-10-01 at 09 31 13

**Pull request merged**